types/index.d.ts

```diff
@@ -0,0 +1,495 @@
+      & (<K extends keyof Registry>(options: RelationshipOptions & { polymorphic: true }) => () => Collection<Registry[K]>);
+
+  /** Declares a one-to-many relationship to another Mirage model type. */
+  export function hasMany<T = ModelInstance>(
+    key?: string,
+    options?: RelationshipOptions
+  ): () => Collection<T>;
+  export function hasMany<T = ModelInstance>(
+    options?: RelationshipOptions
+  ): () => Collection<T>;
+}
+
+declare module "miragejs/-types" {
+  // Captures the result of a `Model.extend()` call
+  const ModelData: unique symbol;
+  interface ModelDefinition<Data = {}> {
+    [ModelData]: Data;
+    extend<NewData>(data: NewData): ModelDefinition<Assign<Data, NewData>>;
+  }
+
+  // Captures the result of a `Factory.extend()` call
+  const FactoryData: unique symbol;
+  interface FactoryDefinition<Data = {}> {
+    [FactoryData]: Data;
+    extend<NewData>(data: NewData): FactoryDefinition<Assign<Data, NewData>>;
+  }
+
+  // The type-level equivalent of `Object.assign`
+  type Assign<T, U> = U & Omit<T, keyof U>;
+
+  // Extract relationship values from a model definiton
+  type FlattenRelationships<T> = {
+    [K in keyof T]: T[K] extends (() => infer Value) ? Value : T[K];
+  };
+
+  // Extract factory method return values from a factory definition
+  type FlattenFactoryMethods<T> = {
+    [K in keyof T]: T[K] extends (n: number) => infer V ? V : T[K];
+  };
+
+  // Extracts model definition info for the given key, if a correesponding model is defined
+  type ExtractModelData<Models, K> = K extends keyof Models
+    ? Models[K] extends ModelDefinition<infer Data>
+      ? FlattenRelationships<Data>
+      : {}
+    : {};
+
+  // Extracts factory definition info for the given key, if a correesponding factory is defined
+  type ExtractFactoryData<Factories, K> = K extends keyof Factories
+    ? Factories[K] extends FactoryDefinition<infer Data>
+      ? FlattenFactoryMethods<Data>
+      : {}
+    : {};
+
+  /**
+   * Models all available information about a given set of model and
+   * factory definitions, determining the behavior of ORM methods on
+   * a `Server` and its corresponding `Schema` instance.
+   */
+  export type Registry<
+    Models extends Record<string, ModelDefinition>,
+    Factories extends Record<string, FactoryDefinition>
+  > = {
+    [K in keyof Models | keyof Factories]: ModelInstance<
+      ExtractModelData<Models, K> & ExtractFactoryData<Factories, K>
+    >;
+  };
+
+  /** Represents the type of an instantiated Mirage model.  */
+  export type ModelInstance<Data = {}> = Data & {
+    id?: string;
+    attrs: Record<string, unknown>;
+    modelName: string;
+
+    /** Persists any updates on this model back to the Mirage database. */
+    save(): void;
+
+    /** Updates and immediately persists a single attr on this model. */
+    update<K extends keyof Data>(key: K, value: Data[K]): void;
```

Comment:
mirage also allows to [update multiple attrs](https://miragejs.com/docs/main-concepts/models#updating-and-deleting-models)

```ts
/** Updates and immediately persists a single or multiple attr(s) on this model. */
update(changes: Partial<Data>): void;
```

Reply:
@NonPolynomial Thanks. I just extended the code with your suggestion.